File: backend/app/schemas/shopping_cart.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel


class ShoppingCartAddIn(BaseModel):
    supermarket_product_id: UUID
    quantity: int = 1
    confirm_replace: bool = False


class ShoppingCartAddOut(BaseModel):
    message: str
    cart_replaced: bool = False


class ShoppingCartSavingsOut(BaseModel):
    total_original_price: Decimal
    total_discount_price: Decimal
    total_lei_saved: Decimal
    currency: str = "RON"


class ShoppingCartItemOut(BaseModel):
    id: UUID
    quantity: int
    supermarket_product_id: UUID

    product_id: UUID | None = None
    product_name: str | None = None

    supermarket_id: UUID | None = None
    supermarket_name: str | None = None

    original_price: Decimal | None = None
    discount_price: Decimal | None = None
    savings_per_unit: Decimal | None = None
    savings_total: Decimal | None = None
    currency: str | None = None
    expiration_date: date | None = None
    stock_quantity: int | None = None

    created_at: datetime
    updated_at: datetime

File: frontend/types/shopping-cart.ts
export type ShoppingCartItem = {
  id: string;
  quantity: number;
  supermarket_product_id: string;
  product_id: string | null;
  product_name: string | null;
  product_image_url: string | null;
  supermarket_id: string | null;
  supermarket_name: string | null;
  original_price: string | null;
  discount_price: string | null;
  savings_per_unit: string | null;
  savings_total: string | null;
  currency: string | null;
  expiration_date: string | null;
  stock_quantity: number | null;
  created_at: string;
  updated_at: string;
};

export type ShoppingCartAddResponse = {
  message: string;
  cart_replaced: boolean;
};

export type ShoppingCartSavings = {
  total_savings: string;
  currency: string;
};

export type ShoppingCartReservation = {
  id: string;
  status: string;
  created_at: string;
  updated_at: string | null;
  items: {
    id: string;
    supermarket_product_id: string;
    quantity: number;
    reserved_price: string;
    currency: string;
  }[];
};
